### ISSUE TITLE
No parameter named "<attr>" for some django columns even with typing correctly set

### ISSUE THREAD

**Issue by bellini666** [microsoft/pyright#2077]
I'm using pyright with coc-pyright on vim, all on their latest versions.

Trying to define a column on django like this:

```python
created_at = models.DateTimeField(
    verbose_name=_("created at"),
    auto_now_add=True,
)
```

Is giving me an error on `auto_now_add` saying that `error: No parameter named "auto_now_add" (reportGeneralTypeIssues)`

Doing a `reveal_type(models.DateTimeField.__init__)` shows me this:

```
Type of "models.DateTimeField.__init__" is "Overload[(self: DateTimeField[datetime], verbose_name: str | bytes | None = ..., name: str | None = ..., auto_now: bool = ..., auto_now_add: bool = ..., primary_key: bool = ..., max_length: int
 | None = ..., unique: bool = ..., blank: bool = ..., null: Literal[False] = ..., db_index: bool = ..., default: Any = ..., editable: bool = ..., auto_created: bool = ..., serialize: bool = ..., choices: Iterable[Tuple[Any, Any] | Tuple[
str, Iterable[Tuple[Any, Any]]]] | None = ..., help_text: str = ..., db_column: str | None = ..., db_tablespace: str | None = ..., validators: Iterable[(*args: Any, **kwargs: Any) -> None] = ..., error_messages: Dict[str, Any] | None = .
..) -> None, (self: DateTimeField[datetime | None], verbose_name: str | bytes | None = ..., name: str | None = ..., auto_now: bool = ..., auto_now_add: bool = ..., primary_key: bool = ..., max_length: int | None = ..., unique: bool = ...
, blank: bool = ..., null: Literal[True] = ..., db_index: bool = ..., default: Any = ..., editable: bool = ..., auto_created: bool = ..., serialize: bool = ..., choices: Iterable[Tuple[Any, Any] | Tuple[str, Iterable[Tuple[Any, Any]]]] |
 None = ..., help_text: str = ..., db_column: str | None = ..., db_tablespace: str | None = ..., validators: Iterable[(*args: Any, **kwargs: Any) -> None] = ..., error_messages: Dict[str, Any] | None = ...) -> None]"     
```

And it is indeed typed as I'm using `django-types` as you an see here: https://github.com/sbdchd/django-types/blob/main/typings/django/db/models/fields/__init__.pyi#L1086

This also happens with max_digits and decimal_places for DecimalField for example

**Comment by erictraut** [comment]
Could you provide a self-contained code snippet? I'm not that familiar with django and am sure where the `models` or `_` symbols come from. 

I know that django does use some non-standard type semantics that cannot be described using the standard type system in Python. I suspect you're hitting that issue. If you provide a complete code sample, I can confirm.

It's unlikely that this is a problem with pyright.

**Comment by bellini666** [comment]
@erictraut sure! Just reproduced with this minimal example. Please tell me if this is enough

1) Create a venv, activate it and `pip install django django-types` on it
2) Create a single file with this content:

```python
from django.db import models

class Foo(models.Model):
    date_field = models.DateTimeField(
        verbose_name="date field",
        auto_now=True,
    )
    decimal_field = models.DecimalField(
        verbose_name="decimal field",
        null=True,
        default=None,
        blank=True,
        max_digits=2,
    )

reveal_type(models.DecimalField)
reveal_type(models.DecimalField.__init__)
```

When running pyright I get this:

```
❯ pyright
No configuration file found.
No pyproject.toml file found.
stubPath /home/bellini/tmp/experiments/django-bug/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Auto-excluding /home/bellini/tmp/experiments/django-bug/.venv
Found 1 source file
/home/bellini/tmp/experiments/django-bug/foobar.py
  /home/bellini/tmp/experiments/django-bug/foobar.py:6:9 - error: No parameter named "auto_now" (reportGeneralTypeIssues)
  /home/bellini/tmp/experiments/django-bug/foobar.py:13:9 - error: No parameter named "max_digits" (reportGeneralTypeIssues)
  /home/bellini/tmp/experiments/django-bug/foobar.py:17:13 - info: Type of "models.DecimalField" is "Type[DecimalField[Unknown]]"
  /home/bellini/tmp/experiments/django-bug/foobar.py:18:13 - info: Type of "models.DecimalField.__init__" is "Overload[(self: DecimalField[Decimal], verbose_name: str | bytes | None = ..., name: str | None = ..., max_digits: int | None = ..., decimal_places: int | None = ..., primary_key: bool = ..., unique: bool = ..., blank: bool = ..., null: Literal[False] = ..., db_index: bool = ..., default: Any = ..., editable: bool = ..., auto_created: bool = ..., serialize: bool = ..., choices: Iterable[Tuple[Any, Any] | Tuple[str, Iterable[Tuple[Any, Any]]]] | None = ..., help_text: str = ..., db_column: str | None = ..., db_tablespace: str | None = ..., validators: Iterable[(*args: Any, **kwargs: Any) -> None] = ..., error_messages: Dict[str, Any] | None = ...) -> None, (self: DecimalField[Decimal | None], verbose_name: str | bytes | None = ..., name: str | None = ..., max_digits: int | None = ..., decimal_places: int | None = ..., primary_key: bool = ..., unique: bool = ..., blank: bool = ..., null: Literal[True] = ..., db_index: bool = ..., default: Any = ..., editable: bool = ..., auto_created: bool = ..., serialize: bool = ..., choices: Iterable[Tuple[Any, Any] | Tuple[str, Iterable[Tuple[Any, Any]]]] | None = ..., help_text: str = ..., db_column: str | None = ..., db_tablespace: str | None = ..., validators: Iterable[(*args: Any, **kwargs: Any) -> None] = ..., error_messages: Dict[str, Any] | None = ...) -> None]"
2 errors, 0 warnings, 2 infos 
Completed in 0.824sec
```

When checking the contents of the .pyi file that was installed by django-types I get this for the decimal for example:

```python
class DecimalField(
    Generic[_DEC],
    Field[Union[str, float, decimal.Decimal, Combinable], decimal.Decimal],
):
    # attributes
    max_digits: int = ...
    decimal_places: int = ...
    @overload
    def __init__(
        self: DecimalField[decimal.Decimal],
        verbose_name: Optional[Union[str, bytes]] = ...,
        name: Optional[str] = ...,
        max_digits: Optional[int] = ...,
        decimal_places: Optional[int] = ...,
        primary_key: bool = ...,
        unique: bool = ...,
        blank: bool = ...,
        null: Literal[False] = ...,
        db_index: bool = ...,
        default: Any = ...,
        editable: bool = ...,
        auto_created: bool = ...,
        serialize: bool = ...,
        choices: Optional[_FieldChoices] = ...,
        help_text: str = ...,
        db_column: Optional[str] = ...,
        db_tablespace: Optional[str] = ...,
        validators: Iterable[_ValidatorCallable] = ...,
        error_messages: Optional[_ErrorMessagesToOverride] = ...,
    ) -> None: ...
    @overload
    def __init__(
        self: DecimalField[Optional[decimal.Decimal]],
        verbose_name: Optional[Union[str, bytes]] = ...,
        name: Optional[str] = ...,
        max_digits: Optional[int] = ...,
        decimal_places: Optional[int] = ...,
        primary_key: bool = ...,
        unique: bool = ...,
        blank: bool = ...,
        null: Literal[True] = ...,
        db_index: bool = ...,
        default: Any = ...,
        editable: bool = ...,
        auto_created: bool = ...,
        serialize: bool = ...,
        choices: Optional[_FieldChoices] = ...,
        help_text: str = ...,
        db_column: Optional[str] = ...,
        db_tablespace: Optional[str] = ...,
        validators: Iterable[_ValidatorCallable] = ...,
        error_messages: Optional[_ErrorMessagesToOverride] = ...,
    ) -> None: ...
    def __get__(self, instance: Any, owner: Any) -> _DEC: ...  # type: ignore [override]
    def __set__(  # type: ignore [override]
        self, instance: Any, value: Union[str, float, Combinable, _DEC]
    ) -> None: ...
```

As you can see, `max_digits` is typed the same way as the other attributes, the only difference is that it is an attribute specific for `DecimalField`, which inherits from `Field` which has the `verbose_name`, `null`, etc for example. I can also see it in the `reveal_type` output.

I know that django does some "magic" stuff that doesn't work nicely here yet, but this is a simple __init__ argument with no magic.

**Comment by erictraut** [comment]
This appears to be a bug in the `django-types` stubs. The `__init__` methods for `DateTimeField ` and `DecimalField` are correct, but the `__new__` method (which is inherited from `Field`) does not have a parameter named `auto_now`, etc., nor does it accept arbitrary keyword arguments (`**kwargs`).

If you add the following parameter to the `Fields.__new__` method within `django-stubs/db/models/fields/__init__.pyi`, the errors will go away.
```python
        **kwargs: Any
```

I don't know if this is the correct fix. You should work with the maintainer of the `django-types` package to determine the desired fix.

**Comment by erictraut** [comment]
Looking at the django source code, the `Field` class doesn't have a `__new__` method, so the type stubs are suspect.

**Comment by bellini666** [comment]
Hi @erictraut . Ok, Just opened an issue there to see what might be going on. Thanks so far!
